state/state.go

```diff
@@ -211,12 +211,59 @@ func ControllerAccess(st *State, tag names.Tag) (permission.UserAccess, error) {
 	return st.UserAccess(tag.(names.UserTag), st.controllerTag)
 }
 
-// RemoveAllModelDocs removes all documents from multi-model
-// collections. The model should be put into a dying state before call
-// this method. Otherwise, there is a race condition in which collections
-// could be added to during or after the running of this method.
-func (st *State) RemoveAllModelDocs() error {
-	err := st.removeAllModelDocs(bson.D{{"life", Dead}})
+// SetDyingModelToDead sets current dying model to dead.
+func (st *State) SetDyingModelToDead() error {
+	model, err := st.Model()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if model.Life() != Dying {
+		return errors.Errorf("can't remove model: model is not dying")
+	}
+	buildTxn := func(attempt int) ([]txn.Op, error) {
+		ops := []txn.Op{{
+			C:      modelsC,
+			Id:     st.ModelUUID(),
+			Assert: isDyingDoc,
+			Update: bson.M{"$set": bson.M{
+				"life":          Dead,
+				"time-of-death": st.nowToTheSecond(),
+			}},
+		}, {
+			// Cleanup the owner:modelName unique key.
+			C:      usermodelnameC,
+			Id:     model.uniqueIndexID(),
+			Remove: true,
+		}}
+		return ops, nil
+	}
+	if err = st.db().Run(buildTxn); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// RemoveModel sets current model to dead then removes all documents from
+// multi-model collections.
+func (st *State) RemoveModel() error {
```

Comment:
I think this should be RemoveDyingModel(). I think that we should never short circuit model.Destroy() to set the next life to Dead. This also improves IAAS model destruction because even for IAAS models Destroy() can take time to run. So in places where we set next model life to Dead because there are no entity refs, perhaps the next life should always be Dying; the undertaker runs environ.Destroy(); then calls RemoveDyingModel(). I think that allows the SetDyingModelToDead()  call to not be needed.